Suggestion for supporting single header fields.

danbev opened the pull request:
Motivation:
At the moment if you want to return a HTTP header containing multiple
values you have to set/add that header once with the values wanted. If
you used set/add with an array/iterable multiple HTTP header fields will
be returned in the response.

Note, that this is indeed a suggestion and additional work and tests
should be added. This is mainly to bring up a discussion.

Modifications:
Added a flag to specify that when multiple values exist for a single
HTTP header then add them as a comma separated string.

Result:
Allows for responses to be smaller.

danbev commented:
@Scottmitch @trustin I've added support for ```add/addObject``` with a single value. I did not add any other methods (like addChar, addByte, addFloat, addLong, addInt) but perhaps I should?

There was a failure in my last commit (3e871fe) where a [leak](https://gist.github.com/danbev/c98cba962170028c7d69) was reported. I'm happy to look into this more closely but just wanted to ask if you recognise it (I did a quick search but did not find anything recent)

Scottmitch commented:
@danbev - Welcome back and I Hope you enjoyed the break!  I'll take a look at your changes in a bit.

>I did not add any other methods (like addChar, addByte, addFloat, addLong, addInt) but perhaps I should?

My initial thoughts...The `add<PrimitiveType>` methods all end up calling `add(..)` and so I don't think we have to override these.

I don't have much context/exposure to this code (looks like `io.netty.handler.codec.compression.IntegrationTest.testEmpty`) but this is the first I am seeing the leak you describe.  Based on the limited information I have seen so far I don't think it would be related to these changes.  Would you mind creating an issue with the information you have gathered thus far to investigate/fix?

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +401,188 @@ public TextHeaders clear() {
+    /*
+     * Will add multiple values for the same header as multiple separate headers.
+     */
+    private class MultipleFieldsValueComposer implements ValuesComposer {
```
Add `final` keyword.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +401,188 @@ public TextHeaders clear() {
+        TextHeaders add(CharSequence name, CharSequence... values);
+        TextHeaders add(CharSequence name, Iterable<? extends CharSequence> values);
+
+        TextHeaders addObject(CharSequence name, Object value);
```
I don't think we need to override this.  All `add*(...)` methods that don't have collection parameters should be calling `add(...)`.  See [DefaultHeaders.java#L504](https://github.com/netty/netty/blob/master/codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java#L504).

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +401,188 @@ public TextHeaders clear() {
+    }
+
+    /*
+     * Will add multiple values for the same header as single header with a comma separated list of values.
```
Perhaps add a reference to the RFC here (https://tools.ietf.org/html/rfc7230#section-3.2.2)?

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +401,188 @@ public TextHeaders clear() {
+            if (values.length > 0) {
+                int end = values.length - 1;
+                for (int i = 0 ; i < end; i++) {
+                    singleValue.append(values[i]).append(", ");
```
I think we need to be care about escaping values.  See https://tools.ietf.org/html/rfc7231#section-8.3.1.

Scottmitch replied:
Also note that all values are not of the native type (CharSequence) are serialized/unserialized by the [ValueConverter](https://github.com/netty/netty/blob/master/codec/src/main/java/io/netty/handler/codec/Headers.java#L54) interface.  I think we may need to add an accessor to be able to use this here.  See [DefaultHeaders.java#L518](https://github.com/netty/netty/blob/master/codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java#L518) for a usage example. I think this would apply to both `commaSeparated` methods.  WDYT?

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +401,188 @@ public TextHeaders clear() {
+            StringBuilder singleValue = new StringBuilder();
+            Iterator<?> iterator = values.iterator();
+            while (iterator.hasNext()) {
+                singleValue.append(iterator.next());
```
I think we need to be care about escaping values.  See https://tools.ietf.org/html/rfc7231#section-8.3.1.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +401,188 @@ public TextHeaders clear() {
+
+        @Override
+        public TextHeaders add(CharSequence name, CharSequence... values) {
+            DefaultTextHeaders.super.add(name, commaSeparated(values));
```
What if a header with identified by `name` already exists?  We may have to do similar logic as you have in the single case.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +401,188 @@ public TextHeaders clear() {
+
+        @Override
+        public TextHeaders add(CharSequence name, Iterable<? extends CharSequence> values) {
+            DefaultTextHeaders.super.add(name, commaSeparated(values));
```
What if a header with identified by `name` already exists?  We may have to do similar logic as you have in the single case.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +401,188 @@ public TextHeaders clear() {
+        }
+
+        @Override
+        public TextHeaders addObject(CharSequence name, Object value) {
```
I think we can get rid of this.  See https://github.com/netty/netty/pull/3271/files#r22608659.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +401,188 @@ public TextHeaders clear() {
+        }
+
+        @Override
+        public TextHeaders addObject(CharSequence name, Object value) {
```
I think we can get rid of this.  See https://github.com/netty/netty/pull/3271/files#r22608659.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -162,6 +163,10 @@ public CharSequence toUnconvertedType(String value) {
 
     private static final NameConverter<CharSequence> CHARSEQUENCE_IDENTITY_CONVERTER =
             new IdentityNameConverter<CharSequence>();
+    // An estimate of the size of a header value.
```
Change to javadoc format?

Scottmitch commented:
@danbev - Done with another round.  Ping me when you are ready again.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +401,188 @@ public TextHeaders clear() {
+                for (int i = 0 ; i < end; i++) {
+                    singleValue.append(values[i]).append(", ");
+                }
+                singleValue.append(values[end]);
```
I think we need to be careful about escaping values.  See https://tools.ietf.org/html/rfc7231#section-8.3.1.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +401,188 @@ public TextHeaders clear() {
+            Iterator<?> iterator = values.iterator();
+            while (iterator.hasNext()) {
+                singleValue.append(iterator.next());
+                if (iterator.hasNext()) {
```
Consider a bit of loop unrolling to removing the duplicate `hasNext()` check in the general case:

```java
        StringBuilder singleValue = new StringBuilder();
        Iterator<?> iterator = values.iterator();
        Object next = iterator.hasNext() ? iterator.next() : null;
        if (next != null) {
            for (Object nextNext = iterator.hasNext() ? iterator.next() : null;
                    nextNext != null;
                    next = nextNext, nextNext = iterator.hasNext() ? iterator.next() : null) {
                singleValue.append(next).append(", ");
            }
            singleValue.append(next);
        }

        return singleValue.toString();
```

It may be a bit more convoluted but I'm interested in your thoughts.

Scottmitch replied:
Note...If we expect/allow `null` values this will prematurely exit in its current form (this can be avoided if needed).

Scottmitch replied:
Here is a simplified version that doesn't have the same limitations w.r.t to `null`:

```java
        StringBuilder singleValue = new StringBuilder();
        Iterator<?> iterator = values.iterator();
        if (iterator.hasNext()) {
            Object next = iterator.next();
            while (iterator.hasNext()) {
                singleValue.append(next).append(", ");
                next = iterator.next();
            }
            singleValue.append(next);
        }
        return singleValue.toString();
```

Note that I have not tested this code :)

danbev commented:
@Scottmitch Thanks, hope you had a good christmas break as well! 

I've create https://github.com/netty/netty/issues/3312 for the CI failure and will take a look at your comments later today. Thanks!

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,205 @@ public TextHeaders clear() {
+
+        private StringBuilder doubleQuoted(StringBuilder singleValue, Object value) {
+            CharSequence converted = valueConverter.convertObject(value);
+            return converted.charAt(0) == DOUBLE_QUOTE ?
```
before you do `converted.charAt(0)` I think you need to check `converted.length() > 0`.

Scottmitch replied:
We may have to do more work when escaping values into a CSV format.  For example see `quoted-string` from https://tools.ietf.org/html/rfc7230#section-3.2.6.  If we have to look the contents of each string I'm wondering if we should also conditionally escape the values (i.e only double-quote and escape if the value contains a `,` character).  WDYT?

danbev replied:
@Scottmitch Yeah, that makes sense. I'll take a look.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,205 @@ public TextHeaders clear() {
+        @Override
+        public TextHeaders add(CharSequence name, CharSequence... values) {
+            add(name, commaSeparated(values));
+            return DefaultTextHeaders.this;
```
nit: Remove this line and just `return` on the line above.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,205 @@ public TextHeaders clear() {
+
+        @Override
+        public TextHeaders add(CharSequence name, Iterable<? extends CharSequence> values) {
+            add(name, commaSeparated(values));
```
nit: Remove this line and just `return` on the line above.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,205 @@ public TextHeaders clear() {
+        @Override
+        public TextHeaders add(CharSequence name, Iterable<? extends CharSequence> values) {
+            add(name, commaSeparated(values));
+            return DefaultTextHeaders.this;
```
nit: Remove this line and just `return` on the line above.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,205 @@ public TextHeaders clear() {
+
+        @Override
+        public TextHeaders add(CharSequence name, CharSequence... values) {
+            add(name, commaSeparated(values));
```
Would it make sense to have a `private TextHeaders add(CharSequence name, CharSequence commaSepeartedValue)` that this and the existing `add(...)` call directly to avoid calling `commaSeparated(...)` twice?

Scottmitch replied:
In case this is the only header identified by `name` they `commaSeparated` may be called twice.  Depending on where we end up with the escape code this may escape the value twice.  Is this a problem, and can we avoid this situation?

danbev replied:
@Scottmitch Will take a look at this together with the escaping you reported above.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,205 @@ public TextHeaders clear() {
+        public TextHeaders add(CharSequence name, CharSequence value) {
+            CharSequence current = DefaultTextHeaders.super.get(name);
+            if (current == null) {
+                DefaultTextHeaders.super.add(name, commaSeparated(value));
```
calling `commaSeparated` here will invoke `valueConverter.convertObject(value)`.  Not that this is necessarily a problem (it depends upon the user's object conversion routine) but it will be the first time the object converted will be explicitly called when we know the type is "native".  WDYT, can we avoid calling `valueConverter.convertObject(value)` in this case?

Scottmitch commented:
@danbev - Finished another round.  I think we are almost there.

danbev commented:
@Scottmitch Thanks! Hope to be able to take them on today but might be Monday by the looks of my current task here.

danbev commented:
@Scottmitch You you mind doing another round of reviews?

Scottmitch commented:
@danbev - Sure.  I'll take a look.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,237 @@ public TextHeaders clear() {
+
+        private StringBuilder quote(StringBuilder sb, CharSequence value) {
+            String converted = value.toString();
+            if (converted.isEmpty()) {
```
nit: I'm not sure you need this empty check.  if `converted.contains(COMMA)` is true then the string will not be empty and `converted.charAt(0)` is OK.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,237 @@ public TextHeaders clear() {
+            return DefaultTextHeaders.this;
+        }
+
+        private TextHeaders addConverted(CharSequence name, CharSequence convertedValue) {
```
It looks like `add` and `addConverted` are the same.  Am I missing something?

danbev replied:
Oh, they are in fact. ```add``` should be delegating to ```addConverted```. Let me update this.

Scottmitch replied:
Oops I moved my comment to https://github.com/netty/netty/pull/3271#discussion-diff-22869993.  Figured it was a better spot to be removed when you updated the code.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,237 @@ public TextHeaders clear() {
+            return sb.toString();
+        }
+
+        private CharSequence commaSeparate(Iterable<? extends CharSequence> values) {
```
I'm trying to think of a way we can avoid the duplication between `commaSeparate` and `convertAndCommaSeparate`. Would it make sense to create a new interface to avoid duplicating this algorithm with `private CharSequence convertAndCommaSeparate(Iterable<?> values)`?  What about something like the following:

```java
private interface CsvValueEscaper<T> {
        StringBuilder escape(StringBuilder input, T value);
    }

    static final CsvValueEscaper<Object> ObjectEscaper = new CsvValueEscaper<Object>() {
        @Override
        public StringBuilder escape(StringBuilder input, Object value) {
            // use the convert method and then quote
        }
    };

    static final CsvValueEscaper<CharSequence> CharEscaper = new CsvValueEscaper<CharSequence>() {
        @Override
        public StringBuilder escape(StringBuilder input, CharSequence value) {
            // just quote
        }
    };

    private static <T> CharSequence commaSeparate(Iterable<? extends T> values, CsvValueEscaper<T> escaper) {
        StringBuilder sb = new StringBuilder();
        Iterator<? extends T> iterator = values.iterator();
        if (iterator.hasNext()) {
            T next = iterator.next();
            while (iterator.hasNext()) {
                escaper.escape(sb, next).append(", ");
                next = iterator.next();
            }
            escaper.escape(sb, next);
        }
        return sb.toString();
    }
```

Then to use the interface something like:

```java
Iterable<? extends CharSequence> values = null;
Iterable<?> values2 = null;
        
commaSeparate(values, CharEscaper);
commaSeparate(values2, ObjectEscaper);
```

Scottmitch replied:
The names of the interfaces are just for demonstration purposes.  I think they can be improved :)

danbev replied:
Sounds good, let me take a stab at this later this evening or tomorrow as I need to head off.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,237 @@ public TextHeaders clear() {
+        public TextHeaders add(CharSequence name, CharSequence value) {
+            CharSequence currentValue = DefaultTextHeaders.super.get(name);
+            if (currentValue == null) {
+                DefaultTextHeaders.super.add(name, value);
```
It looks like `add` and `addConverted` are the same. Am I missing something?

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,237 @@ public TextHeaders clear() {
+        }
+
+        private CharSequence commaSeparate(CharSequence currentValue, CharSequence value) {
+            return new StringBuilder().append(currentValue).append(", ").append(value).toString();
```
We should be able to provide an exact initial size. Something like `new StringBuilder(currentValue.length() + 2 + value.length())`

Scottmitch replied:
I think we should first try to consolidate `commaSeparate` and `convertAndCommaSeparate` as suggested in https://github.com/netty/netty/pull/3271/files#r22869589, but I think we may need to check for escaping `value`.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,228 @@ public TextHeaders clear() {
+        }
+
+        private StringBuilder quote(StringBuilder sb, CharSequence value) {
+            String converted = value.toString();
```
In general using `String` methods is a good idea but I'm wondering if we need to do that here.  Can we avoid the `toString()` and just do a loop (and break out if we find the `,` character)?  This way we won't need the `COMMA` static either.

Scottmitch commented:
@danbev - I'm done...Back to you :)

danbev commented:
@Scottmitch Thanks again! I've updated with two commits to address your latest comments.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,226 @@ public TextHeaders clear() {
+    private final class SingleHeaderValuesComposer implements ValuesComposer {
+
+        private final ValueConverter<CharSequence> valueConverter;
+        private final ValueEscaper<Object> objectEscaper = new ValueEscaper<Object>() {
```
Add `static` keyword.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,226 @@ public TextHeaders clear() {
+                return quote(input, valueConverter.convertObject(value));
+            }
+        };
+        private final ValueEscaper<CharSequence> charSequenceEscaper = new ValueEscaper<CharSequence>() {
```
Add `static` keyword.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,226 @@ public TextHeaders clear() {
+        }
+    }
+
+    private interface ValueEscaper<T> {
```
Add comments

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,226 @@ public TextHeaders clear() {
+    }
+
+    private interface ValueEscaper<T> {
+        StringBuilder escape(StringBuilder input, T value);
```
Add comments

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,226 @@ public TextHeaders clear() {
+        };
+
+        SingleHeaderValuesComposer() {
+            valueConverter = valueConverter();
```
Now that `objectEscaper` is initialized inline and using the `valueConveter` consider making the initialization consistent. Either initialize them both in the constructor or initialize them both inline.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,226 @@ public TextHeaders clear() {
+        }
+
+        private CharSequence escape(CharSequence value) {
+            StringBuilder sb = new StringBuilder(value.length() + 4);
```
Perhaps add a comment as to why 4 was chosen.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,226 @@ public TextHeaders clear() {
+                    .append(value).toString();
+        }
+
+        private StringBuilder quote(StringBuilder sb, CharSequence value) {
```
I'm not sure that quoting and escaping are divisible operations.  Perhaps clarification of method names will help but I think this algorithm may some updates.  Here is a summary of what I think the two terms mean:

- Escape: Special characters need to be preceded with a "token" to delineate the following character is content, and not a control character.
- Quote: If a special character is found, and the value has not already been Escaped and Quoted, then the value needs to be wrapped in `"` characters.

See http://grepcode.com/file/repo1.maven.org/maven2/org.apache.commons/commons-lang3/3.1/org/apache/commons/lang3/StringEscapeUtils.java#StringEscapeUtils.escapeCsv%28java.lang.String%29 for a reference of what we likely need to do.  The comments on this method are also helpful.

Scottmitch replied:
I'm not sure if Netty has a CSV Escape method yet, but it may be good to make one more generally available as part of this effort.

Scottmitch replied:
We should also add a unit test that adds the same value again and check that the result is the same:

```java
headers.add("foo", "some, value, with stuff \"");
String value = headers.get("foo");
// Possibly do the following 2 lines in a loop
headers.add("foo", value);
assertEquals(value, headers.get("foo"));
```

Scottmitch commented:
@danbev - Finished another round.

Scottmitch commented:
We should also add a unit test that adds the same value again and check that the result is the same:

headers.add("foo", "some, value, with stuff \""); // this test can be repeated with other interesting inputs (i.e. empty, ",", "\"", etc...)
String value = headers.get("foo");
for (int i = 0; i < /*some number*/; ++i) {
  headers.add("foo", value);
  assertEquals(value, headers.get("foo"));
}

danbev commented:
I'll have some time tomorrow and will take a look this and the other issues
you reported.

torsdag 15 januari 2015 skrev Scott Mitchell <notifications@github.com>:

> We should also add a unit test that adds the same value again and check
> that the result is the same:
>
> headers.add("foo", "some, value, with stuff \""); // this test can be
> repeated with other interesting inputs (i.e. empty, ",", "\"", etc...)
> String value = headers.get("foo");
> for (int i = 0; i < /*some number*/; ++i) {
> headers.add("foo", value);
> assertEquals(value, headers.get("foo"));
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/netty/netty/pull/3271#issuecomment-70096483>.
>

danbev commented:
@Scottmitch Sorry about the delay on this. Still not happy with this but have run out of time today. Will revisit on Monday.

Scottmitch commented:
@danbev - No problem at all.  Keep up the good work.  Should I look again now, or wait?

danbev commented:
@Scottmitch I've extracted the quoting of a ```quoteCsv``` method in [StringUtil](https://github.com/danbev/netty/blob/single-http-header-fields/common/src/main/java/io/netty/util/internal/StringUtil.java#L339). 
I've also updated DefaultTextHeaders to use the above method, and changed the naming from escape to quote where applicable. 

Unless I'm mistaken, there is no escaping at the moment for values added using the default (multiple header fields) implementation. Should we leave this up to the developer to make sure that any values are escaped properly?

I'm still a little confused about the terminology here. I think your definitions [above](https://github.com/netty/netty/pull/3271#discussion_r22950120) make sense, but when reading the [javadocs](http://grepcode.com/file/repo1.maven.org/maven2/org.apache.commons/commons-lang3/3.1/org/apache/commons/lang3/StringEscapeUtils.java#StringEscapeUtils.escapeCsv%28java.lang.String%29) you referred to they mention _escape_ and not _quote_.

Scottmitch commented:
> Unless I'm mistaken, there is no escaping at the moment for values added using the default (multiple header fields) implementation.

You are not mistaken.  There should be no escaping necessary because we are not forcing CSV, but instead separated by duplicating the header name with a new value.

> I'm still a little confused about the terminology here. I think your definitions above make sense, but when reading the javadocs you referred to they mention escape and not quote.

I think we should use the `escape` terminology from an API perspective.  The quote operation is part of escaping for CSV.  The intention of providing the definitions was to distinguish the two operations and clarify their relationship.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,217 @@ public TextHeaders clear() {
+    private final class SingleHeaderValuesComposer implements ValuesComposer {
+
+        private final ValueConverter<CharSequence> valueConverter = valueConverter();
+        private final CsvValueQuoter<Object> objectQuoter = new CsvValueQuoter<Object>() {
```
Lets use `escape` instead of `quote`.  I think this is a global comment for this PR.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,217 @@ public TextHeaders clear() {
+            if (currentValue == null) {
+                DefaultTextHeaders.super.add(name, escapedValue);
+            } else {
+                if (!currentValue.equals(escapedValue)) {
```
Is this correct?  Consider the following example:

```java
headers.add("foo", "goo");
headers.add("foo", "goo");
```

Shouldn't the result be `foo: goo, goo`?

It seems like we should not be filtering additions based upon the value.  Should we add a unit test for this?

danbev replied:
Will take a look at this today.

Scottmitch commented on common/src/test/java/io/netty/util/internal/StringUtilTest.java:
```diff
@@ -81,4 +81,61 @@ public void splitMaxPart() {
+
+    @Test
+    public void quoteCsvEmpty() {
+        assertEquals("", quoteCsv(""));
```
Lets make a utility method in this class to run all test cases through a loop.

```java
private static void testInLoop(String expected, String preEscaped) {
  String escapedValue = preEscaped;
  for (int i = 0; i < /*some number*/; ++i) {
    escapedValue = quoteCsv(escapedValue);
    assertEquals(expected, escapedValue);
  }
```

danbev replied:
Sounds good, I'll add this.

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -314,6 +321,50 @@ public static String simpleClassName(Class<?> clazz) {
+     * @param value the value to be appended, and quoted if necessary
+     * @return {@link CharSequence} the quoted value
+     */
+    public static CharSequence quoteCsv(CharSequence value) {
```
I think we need a function which is `escapeCsv`.  This function should do the following:

1. Escape all special characters that are not escaped (precede with `"` character).
2. Wrap value in quotes if any escapes where added or `,` character is found (and value is not already wrapped in quotes).

danbev replied:
Let me take a stab at this later today.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +404,217 @@ public TextHeaders clear() {
+            if (values.length > 0) {
+                int end = values.length - 1;
+                for (int i = 0; i < end; i++) {
+                    sb.append(quoter.quote(values[i])).append(", ");
```
I think we should just be appending a `,` character (no additional space).  WDYT?

Scottmitch replied:
If you agree there are 2 other places where this exists below.

danbev replied:
Yep, I agree. Will fix this as well.

danbev commented:
@Scottmitch I've modified the [escapeCsv](https://github.com/danbev/netty/blob/single-http-header-fields/common/src/main/java/io/netty/util/internal/StringUtil.java#L325) method and it is followings what is done [here](http://grepcode.com/file/repo1.maven.org/maven2/org.apache.commons/commons-lang3/3.1/org/apache/commons/lang3/StringEscapeUtils.java#StringEscapeUtils.0ESCAPE_CSV). 
Basically, if a CharSequence contains a comma character, a line feed character, carriage return character, or a double quote character, the whole CharSequence is enclosed in double quotes. In addition, if the CharSequence contained a double quote character it is escaped with a double quote character as described in [RFC-4180](https://tools.ietf.org/html/rfc4180#section-2). Is this what you had in mind when you wrote https://github.com/netty/netty/pull/3271#discussion_r23261233

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +405,214 @@ public TextHeaders clear() {
+    private final class SingleHeaderValuesComposer implements ValuesComposer {
+
+        private final ValueConverter<CharSequence> valueConverter = valueConverter();
+        private final CsvValueEscaper<Object> objectEscaper = new CsvValueEscaper<Object>() {
```
I'm wondering if it makes sense to lazy initialize this.  Depending on the use case it may not be used.  WDTY?

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +405,214 @@ public TextHeaders clear() {
+        }
+
+        private CharSequence commaSeparateEscapedValues(CharSequence currentValue, CharSequence value) {
+            return new StringBuilder(currentValue.length() + 2 + value.length())
```
`+ 2` -> `+ 1`?

danbev replied:
Yes. Self face slap.

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -314,6 +322,67 @@ public static String simpleClassName(Class<?> clazz) {
+        if (length == 0) {
+            return value;
+        }
+        int first = 0;
```
is this needed?  Can we just use `0` instead?

danbev replied:
Not really, just had it there to be consistent later in the switch clause. Let me remove it.

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -314,6 +322,67 @@ public static String simpleClassName(Class<?> clazz) {
+        int last = length - 1;
+        boolean quoted = isDoubleQuote(previous) && isDoubleQuote(value.charAt(last)) && length != 1;
+        boolean needsQuote = false;
+        StringBuilder escaped = new StringBuilder(DOUBLE_QUOTE_STRING);
```
Would an initial size of `value.length()` make more sense?  Possibly plus some delta size to anticipate quoting and escaping?

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -314,6 +322,67 @@ public static String simpleClassName(Class<?> clazz) {
+                    // if the current char is not escaped, escape it. But only if it is not already escaped.
+                    // special case where a double quote exists before the last closing quote.
+                    if (!quoted || !isDoubleQuote(previous) &&
+                            (!isDoubleQuote(next(value, i)) || isDoubleQuote(next(value, i)) && i + 1 == last)) {
```
Is it possible for `i == length -1` (aka `i == last`)?  Seems like `next(...)` will go out of bounds in this case.

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -314,6 +322,67 @@ public static String simpleClassName(Class<?> clazz) {
+                case LINE_FEED:
+                case CARRIAGE_RETURN:
+                case COMMA:
+                    needsQuote = true;
```
Shouldn't we be checking if all special characters need to be escaped?  Seems like we are only checking to escape `DOUBLE_QUOTE`?

Scottmitch replied:
Never mind.  https://tools.ietf.org/html/rfc4180#section-2 clarifies the approach.

danbev replied:
Sorry about being a little slow here, but which special characters are you referring to? Are these the ones from [rfc2616](http://www.w3.org/Protocols/rfc2616/rfc2616-sec2.html#sec2.2)?

Scottmitch replied:
Nope your not slow.  Just have seen so many different CSV escape approaches.  Having the rfc 4180 in the comments clarified things.

Scottmitch replied:
However what if we have the following input:

`"foo,goo"`

Then you will evaluate `needsQuote = true` unconditionally.  What if `needsQuote` is changed to `foundSpecialCharacter` and then at the end you check `if foundSpecialCharacter && !quoted ? /*escaped value*/ : value`?

danbev replied:
I've updated this but I think we need an additional condition here. If a string is already quoted, and contains a single double quote:
```"foo"goo"```

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -314,6 +322,67 @@ public static String simpleClassName(Class<?> clazz) {
+        return c == DOUBLE_QUOTE;
+    }
+
+    private static char next(CharSequence charSequence, int index) {
```
Instead of a method, lets store the result of `charSequence.charAt(index + 1);` in a variable and use that.  This method is used twice in the same code block.

danbev replied:
Will fix

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -314,6 +322,67 @@ public static String simpleClassName(Class<?> clazz) {
+            return value;
+        }
+        int first = 0;
+        char previous = value.charAt(first);
```
This variable is only needed a small percentage of the time (only when `"` is found).  Consider removing this variable and just doing a `value.charAt(i-1)` (with a `if(i != 0)` conditional protecting this) when needed.

danbev replied:
Will fix.

Scottmitch commented on common/src/test/java/io/netty/util/internal/StringUtilTest.java:
```diff
@@ -81,4 +81,165 @@ public void splitMaxPart() {
+        CharSequence escapedValue = value;
+        for (int i = 0; i < 10; ++i) {
+            escapedValue = escapeCsv(escapedValue);
+            assertThat(escapedValue, equalTo(expected));
```
Use `assertEquals`

Scottmitch commented on common/src/test/java/io/netty/util/internal/StringUtilTest.java:
```diff
@@ -81,4 +81,165 @@ public void splitMaxPart() {
+        assertEscaped(value, expected);
+    }
+
+    private static void assertEscaped(CharSequence value, CharSequence expected) {
```
consider renaming to avoid potential confusion with JUNIT assert* methods.

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -314,6 +322,67 @@ public static String simpleClassName(Class<?> clazz) {
+                    }
+                    // if the current char is not escaped, escape it. But only if it is not already escaped.
+                    // special case where a double quote exists before the last closing quote.
+                    if (!quoted || !isDoubleQuote(previous) &&
```
Not sure if this is a problem but may cause confusion...`previous` may be equal to the `current` if `i==0 && !quoted`.  I think this suggestion https://github.com/netty/netty/pull/3271/files#r23380380 will help clarify things.

danbev replied:
Will fix.

Scottmitch commented:
@danbev - After we nail down the `escapeCsv` we should contribute an `unescapeCsv` method.  We can also discuss if it makes sense to modify the `get*` methods in `SingleHeaderValuesComposer` to use this method to unescape values.

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -314,6 +322,67 @@ public static String simpleClassName(Class<?> clazz) {
+                        continue;
+                    }
+                    // last charater but value is not quoted, so it needs to be escaped.
+                    if (i == last) {
```
To reduce the conditional blocks we could roll the `if` check above (i.e. `if ((i == 0 || i == last) && quoted)`) into this if/else block.  WDTY?

```java
if (i == last) {
  if (quoted) continue;
  escaped.append(DOUBLE_QUOTE);
} else if (i == 0 && quoted) {
  continue;
} else {
...
}
```

Scottmitch replied:
Note that I have edited the above code since it was originally posted :)

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -314,6 +322,67 @@ public static String simpleClassName(Class<?> clazz) {
+        boolean quoted = isDoubleQuote(value.charAt(0)) && isDoubleQuote(value.charAt(last)) && length != 1;
+        boolean foundSpecialCharacter = false;
+        boolean foundEscapedCharacter = false;
+        StringBuilder escaped = new StringBuilder(length).append(DOUBLE_QUOTE_STRING);
```
At a minimum I think we need `length + 2`.  We may also want to add some other small delta to account for escape characters that may be added (this would be a guess).

Scottmitch replied:
also change `DOUBLE_QUOTE_STRING` to `DOUBLE_QUOTE`.

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -30,6 +33,11 @@
+    public static final char COMMA = ',';
+    public static final char LINE_FEED = '\n';
+    public static final char CARRIAGE_RETURN = '\r';
+    public static final String DOUBLE_QUOTE_STRING = String.valueOf(DOUBLE_QUOTE);
```
I think this is only needed in the test case `DefaultTextHeadersTest`.  Can we move it there?

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -314,6 +322,67 @@ public static String simpleClassName(Class<?> clazz) {
+        int last = length - 1;
+        boolean quoted = isDoubleQuote(value.charAt(0)) && isDoubleQuote(value.charAt(last)) && length != 1;
+        boolean foundSpecialCharacter = false;
+        boolean foundEscapedCharacter = false;
```
`foundEscapedCharacter` -> `foundEscapeCharacter`.  Remove the `d` to clarify meaning.  You are not tracking if you found a character that was already `escaped`, but instead you are tracking if you found an `escape` character.

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -314,6 +322,67 @@ public static String simpleClassName(Class<?> clazz) {
+                        escaped.append(DOUBLE_QUOTE);
+                    } else {
+                        boolean isNextCharDoubleQuote = isDoubleQuote(value.charAt(i + 1));
+                        if (!isDoubleQuote(i != 0 ? value.charAt(i - 1) : 0) &&
```
What is the expected result if your input string is `"f`?  Wouldn't it be `"""f"`?  Does this test case pass?  Either way lets add this as a test case.

Scottmitch replied:
Lets also add the following test case:

input: `""f` expected: `"""""f"`

danbev replied:
Yep, that is the expected result. I thought I had a test case for that but don't. Added one (not pushed) yet and it passes.

I've run out of time but hope to have some time to make the other changes during the weekend, otherwise on Monday. Thanks for yet another round of reviews!

Scottmitch replied:
Can we clarify the `!isDoubleQuote(i != 0 ? value.charAt(i - 1) : 0)` conditional a bit?  `0` is being passed into the `isDoubleQuote` function as a character (we shouldn't have to call the method at all in this case, and `0` as a character seems a bit `c/c++` ish :) ).  How about:

`(i != 0 && !isDoubleQuote(value.charAt(i - 1)) && ...`

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -314,6 +322,67 @@ public static String simpleClassName(Class<?> clazz) {
+            escaped.append(current);
+        }
+        return foundEscapedCharacter || (foundSpecialCharacter && !quoted) ?
+                escaped.append(DOUBLE_QUOTE).toString() : value;
```
The return type is `CharSequence`.  Is there a reason why we need to do the `toString()`?

danbev replied:
There is not good reason for calling ```toString()```. I'll remove the call.

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -314,6 +322,67 @@ public static String simpleClassName(Class<?> clazz) {
+     * If a double quote is found inside of the value, it will be escaped by preceeding it
+     * with another double quote.
+     *
+     * @param value the value to be appended, and quoted if necessary
```
This reads a bit awkward.  Could we clarify?  How about:

`@param value The value which will be escaped according to <a href="https://tools.ietf.org/html/rfc4180#section-2">RFC-4180</a>`

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -314,6 +322,67 @@ public static String simpleClassName(Class<?> clazz) {
+     * Escapes the specified value, if necessary according to
+     * <a href="https://tools.ietf.org/html/rfc4180#section-2">RFC-4180</a>.
+     * <p>
+     * The {@link CharSequence} will be escaped is if contains a
```
`{@link CharSequence}` -> `{@code value}`

Scottmitch replied:
On second thought lets either remove this block (the RFC defines the escaping rules) or clarify a bit:

```
The special characters are:
<ul>
...
</ul>
```

The value may not be "escaped" just because these characters are found (if it is already properly quoted and escaped nothing will be done).

Scottmitch commented:
@danbev - Back to you.  The `escapeCsv` method is looking good!

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +405,228 @@ public TextHeaders clear() {
+            return new StringBuilder(currentValue.length() + 1 + value.length())
+                    .append(currentValue)
+                    .append(COMMA)
+                    .append(value).toString();
```
is `toString()` needed?

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -30,6 +33,11 @@
+    public static final char COMMA = ',';
+    public static final char LINE_FEED = '\n';
+    public static final char CARRIAGE_RETURN = '\r';
+    private static final int DEFAULT_NR_OF_DOUBLE_QUOTES = 3;
```
nit: Group the private accessible members together.

Scottmitch replied:
Is `NR` an abbreviation for `NUMBER`?  If so how about the following:

```java
/**
 * 2 = Quotes at beginning and end.
 * 5 = Extra allowance for anticipated escape characters that may be added.
 */
private static final int CSV_NUMBER_ESCAPE_CHARACTERS = 2 + 5;
```

Scottmitch commented on common/src/test/java/io/netty/util/internal/StringUtilTest.java:
```diff
@@ -81,4 +81,186 @@ public void splitMaxPart() {
+        CharSequence escapedValue = value;
+        for (int i = 0; i < 10; ++i) {
+            escapedValue = StringUtil.escapeCsv(escapedValue);
+            assertEquals(expected, escapedValue.toString());
```
is `toString()` needed?

Scottmitch replied:
So it is needed, or was this missed?

danbev replied:
This one is actually needed as ```assertEquals``` will check that both arguments are instances of String.

Scottmitch replied:
sgtm

Scottmitch commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -314,6 +322,57 @@ public static String simpleClassName(Class<?> clazz) {
+            }
+            escaped.append(current);
+        }
+        return foundEscapeCharacter || (foundSpecialCharacter && !quoted) ?
```
`foundEscapeCharacter` is set to true for any `DOUBLE_QUOTE` character that is found.  Is it possible that a string that is already properly escaped would get an additional `"` at the beginning and end?

Scottmitch replied:
Should `foundEscapeCharacter` be `addedEscapeCharacter` and be set to true if a `escaped.append(DOUBLE_QUOTE);` was made?  If this is the case we may not need the variable at all...could we just check `value.length() != escaped.length() - 1` ( `-1` is for the quote appended in the beginning).

danbev replied:
With regards to the first issue if it is possible for an already quoted value to get an additional ```"```` at the beginning and end, this use case is covered by [escapeCsvAlreadyQuoted](https://github.com/danbev/netty/blob/single-http-header-fields/common/src/test/java/io/netty/util/internal/StringUtilTest.java#L105)

Will take a closer look at your second comment and try it out.

Scottmitch replied:
> this use case is covered by escapeCsvAlreadyQuoted

I was thinking if the input is a properly escaped string with a quote character internally (not first or last character).  For example:

input: `"foo""goo"`
output: `"foo""goo"`

Perhaps I'm missing something but seems like just because you find a `"` doesn't mean that you have to add quotes around the value.

danbev replied:
>Perhaps I'm missing something but seems like just because you find a " doesn't mean that you >have to add quotes around the value?

You are right that we should not add quotes around an already quoted value, but we do check if the character before and after the current ```"```, and don't add it unless it is required. I've added a test with the input and expected output that you provided (not pushed yet).

Scottmitch replied:
> but we do check if the character before and after the current ", and don't add it unless it is required

Understood I think the escaping `"` characters looks good.  I am still a bit uncertain about the condition which wraps the entire value in quotes. The `foundEscapeCharacter` is set unconditionally if a `"` is found.  `foundEscapeCharacter` is used here to determine if the "quoted" version of the input should be used. Using this logic wouldn't it be possible to find a string with a `"` that is properly escaped and quoted, but then this code will still wrap the string in additional `"` characters?  Is this line the same in the un-pushed code?

danbev replied:
Sorry for going silent, have another task here at work which I need to give
higher priority. Will circle back as soon as I can.

tisdag 27 januari 2015 skrev Scott Mitchell <notifications@github.com>:

> In common/src/main/java/io/netty/util/internal/StringUtil.java
> <https://github.com/netty/netty/pull/3271#discussion_r23604716>:
>
> > +                    } else {
> > +                        boolean isNextCharDoubleQuote = isDoubleQuote(value.charAt(i + 1));
> > +                        if (!isDoubleQuote(value.charAt(i - 1)) &&
> > +                                (!isNextCharDoubleQuote || isNextCharDoubleQuote && i + 1 == last)) {
> > +                            escaped.append(DOUBLE_QUOTE);
> > +                        }
> > +                    }
> > +                    foundEscapeCharacter = true;
> > +                case LINE_FEED:
> > +                case CARRIAGE_RETURN:
> > +                case COMMA:
> > +                    foundSpecialCharacter = true;
> > +            }
> > +            escaped.append(current);
> > +        }
> > +        return foundEscapeCharacter || (foundSpecialCharacter && !quoted) ?
>
>  but we do check if the character before and after the current ", and
> don't add it unless it is required
>
> Understood I think the escaping " characters looks good. I am still a bit
> uncertain about the condition which wraps the entire value in quotes. The
> foundEscapeCharacter is set unconditionally if a " is found.
> foundEscapeCharacter is used here to determine if the "quoted" version of
> the input should be used. Using this logic wouldn't it be possible to find
> a string with a " that is properly escaped and quoted, but then this code
> will still wrap the string in additional " characters? Is this line the
> same in the un-pushed code?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/netty/netty/pull/3271/files#r23604716>.
>

Scottmitch replied:
@danbev - No problem.  Just ping me when you are ready for another round of reviews.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +405,228 @@ public TextHeaders clear() {
+                }
+                sb.append(escaper.escape(next));
+            }
+            return sb.toString();
```
is `toString()` needed?

danbev replied:
Sorry, I missed these. Thanks for pointing them out. Will remove them.

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +405,228 @@ public TextHeaders clear() {
+                }
+                sb.append(escaper.escape(values[end]));
+            }
+            return sb.toString();
```
is `toString()` needed?

slandelle commented on codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java:
```diff
@@ -396,4 +405,228 @@ public TextHeaders clear() {
+        }
+
+        private <T> CharSequence commaSeparate(CsvValueEscaper<T> escaper, T... values) {
+            StringBuilder sb = new StringBuilder(values.length * DEFAULT_VALUE_SIZE);
```
Why create new StringBuilders instead of using pooled `InternalThreadLocalMap#stringBuilder()`?

danbev replied:
I was not aware of ```InternalThreadLocalMap#stringBuilder()```.

Scottmitch replied:
If we are returning the StringBuilder then is [InternalThreadLocalMap.stringBuilder()](https://github.com/netty/netty/blob/master/common/src/main/java/io/netty/util/internal/InternalThreadLocalMap.java#L175) a good idea?  Seems like the value could be overridden/corrupted the next time some other method uses the same accessor.

danbev replied:
Yeah, I noticed the same thing while looking at the method so it might not be a good fit in this case.

Scottmitch commented:
@danbev - Is this still on your radar?

danbev commented:
@Scottmitch It is still on my radar. I've just not had time to revisit this. Will try to make some time for this later this afternoon (have been hoping to do so most of the week but other things have come up).

danbev commented:
@Scottmitch I've add a commit for the issue you brought up in https://github.com/netty/netty/pull/3271#discussion_r23604716. Would you mind reviewing?

Scottmitch commented:
@danbev - Looking good, just 1 outstanding comment!  Can you address the build failure?

Does it make sense to cherry-pick this without the un-escape portion of this (csv-unescape and TextHeaders.get(..) methods which will unescape)?

danbev commented:
@Scottmitch Absolutely, I'll rebase and run the test suite locally and sort out the build failure.

>Does it make sense to cherry-pick this without the un-escape portion of this (csv-unescape and TextHeaders.get(..) methods which will unescape)?

Yeah, I think that would be fine. I'll add an issue for this so that it is not forgotten.

Scottmitch commented:
@danbev Great work, and thanks for sticking with this one.  Please reference this PR from your issue about un-escaping because I think that is key after introducing this functionality.  Are you planning on providing that functionality in the near term future? Feel free to cherry pick (into 4.1 and 5.0).

danbev commented:
@Scottmitch Thanks for all the reviews! 
I've create #3435 for the un-escaping. I hope to be able to handle this though time is running out (I go on parental leave on March 2nd) and I also have a few things at work that need finishing. I'll see how I manage and I'll un-assign it if I can't do it in time.

danbev commented:
Cherry picked into 4.1 as c53b8d5a855468a890757b4f0612d7ae192a3592
Cherry picked into master as c6e88b00a3cc6cf438f74b1a9077191b660df711
